### PR TITLE
feature(msgpack.rpc): introduce vocab for msgpack rpc

### PR DESCRIPTION
**Pull request by rgrinberg**
Implemented according to this spec https://github.com/msgpack-rpc/msgpack-rpc/blob/master/spec.md

I haven't yet written the docs because I'm still working on the API. It's at a point where it's usable though.

### CODE REVIEW

File: basis/msgpack/rpc/rpc.factor [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+TUPLE: request msgid method params ;
+
+TUPLE: response msgid error result ;
+
+TUPLE: notification method params ;
+
+TUPLE: session active? incoming-requests outgoing-requests notification-callback request-callback ;
+
+M: request write-msgpack
+    [ msgid>> ] [ method>> ] [ params>> ]
```
Comment (mrjbq7) [inline basis/msgpack/rpc/rpc.factor:16]: somewhat inefficient, this is faster:

```factor
M: request write-msgpack
     0 swap [ msgid>> ] [ method>> ] [ params>> ] tri 4array ;
```

File: basis/msgpack/rpc/rpc.factor [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+GENERIC: stop ( session -- )
+
+! TODO close active requests
+M: session stop ( session -- )
+    f swap active?<< ;
+
+: <session> ( -- session )
+    t H{ } clone H{ } clone [ drop ] [ drop ] session boa ;
+
+: <request> ( msgid method params -- request )
```
Comment (mrjbq7) [inline basis/msgpack/rpc/rpc.factor:40]: If you aren't going to do anything different, these are auto-generated using constructor syntax:

```
C: <request> request
```

File: basis/msgpack/rpc/rpc.factor [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+: <response-error> ( id error -- response )
+    +msgpack-nil+ response boa ;
+
+M: notification feed-packet
+    swap notification-callback>> call( notification -- ) ;
+
+ERROR: invalid-packet packet ;
+
+: check-array ( obj -- obj )
```
Comment (mrjbq7) [inline basis/msgpack/rpc/rpc.factor:57]: We also have ``array check-instance``

**Comment by mrjbq7** [comment]
Pretty cool! Made a few small comments.

I would probably use ``<PRIVATE`` to keep some of the common words like ``start``, ``stop``, and ``recv-loop`` from shadowing other vocabs.

**Comment by mrjbq7** [comment]
Either way, let me know!

File: basis/msgpack/rpc/rpc.factor (hunk 3 above)
Comment (rgrinberg) [reply basis/msgpack/rpc/rpc.factor:57]: That gives a worse error though

**Comment by rgrinberg** [comment]
I made a bunch of stop private (but not start/stop). Should be ready from my end

**Pull request merged**